# Doctor and Veterinary Classification using NLP

This notebook is for building a model which will correctly classify a number of given reddit users as practicing doctors, practicng veterinary or others based on each user's comments 

The dataset for this task would be sourced from a databased whose link is given as

[postgresql://niphemi.oyewole:W7bHIgaN1ejh@ep-delicate-river-a5cq94ee-pooler.us-east-2.aws.neon.tech/Vetassist?statusColor=F8F8F8&env=&name=redditors%20db&tLSMode=0&usePrivateKey=false&safeModeLevel=0&advancedSafeModeLevel=0&driverVersion=0&lazyload=false](postgresql://niphemi.oyewole:W7bHIgaN1ejh@ep-delicate-river-a5cq94ee-pooler.us-east-2.aws.neon.tech/Vetassist?statusColor=F8F8F8&env=&name=redditors%20db&tLSMode=0&usePrivateKey=false&safeModeLevel=0&advancedSafeModeLevel=0&driverVersion=0&lazyload=false)

However, trying to access the database with the given link would result in errors

Therefore, a modified version of the link would be used

## Module Importations and Data Retrieval

Before continuing, needed libraries would be imported below

In [1]:
import re             # for regrex operations
import string         # for removing punctuations
import numpy as np    # for mathematical calculations
import pandas as pd    # for working with structured data (dataframes)
from sqlalchemy import create_engine # for connecting to database

The modified link to access the database is defined below

In [2]:
# define the connection link
conn_str = "postgresql://niphemi.oyewole:endpoint=ep-delicate-river-a5cq94ee-pooler;W7bHIgaN1ejh@ep-delicate-river-a5cq94ee-pooler.us-east-2.aws.neon.tech/Vetassist?sslmode=require"

# create connection to the databse
engine =  create_engine(conn_str)

First, lets take a look at the tables in the database

In [3]:
# define sql query for retrieving the tables in the database
sql_for_tables = """
SELECT
    table_schema || '.' || table_name
FROM
    information_schema.tables
WHERE
    table_type = 'BASE TABLE'
AND
    table_schema NOT IN ('pg_catalog', 'information_schema');
"""

In [4]:
# retrieve the tables in a dataframe
tables_df = pd.read_sql_query(sql_for_tables, engine)

In [5]:
tables_df

?column?
0  public.reddit_usernames_comments
1           public.reddit_usernames

There are two tables in the database as shown above

Each table would be saved in a pandas dataframe

In [6]:
sql_for_table1 = """
SELECT
    *
FROM
    public.reddit_usernames_comments;
"""

> Note: The code below may take a while to run. If it fails, reconnect the engine above then rerun the cell

In [7]:
user_comment_df = pd.read_sql_query(sql_for_table1, engine)

In [8]:
sql_for_table2 = """
SELECT
    *
FROM
    public.reddit_usernames;
"""

> Note: The code below may take a while to run. If it fails, reconnect the engine above then rerun the cell

In [9]:
user_info_df = pd.read_sql_query(sql_for_table2, engine)

Lets take a look at the tables one after the other

In [10]:
user_comment_df.head()

username                                           comments
0        LoveAGoodTwist  Female, Kentucky.  4 years out. Work equine on...
1            wahznooski  As a woman of reproductive age, fuck Texas|As ...
2  Churro_The_fish_Girl  what makes you want to become a vet?|what make...
3              abarthch  I see of course there are changing variables, ...
4            VoodooKing  I have 412+ and faced issues because wireguard...

In [11]:
user_comment_df.shape

(3276, 2)

## Data Exploration

This table (now dataframe) contains usernames of users and their comments

Lets look at a comment in order to understand how it is structured

In [12]:
# print all comments for first user
user_comment_df["comments"][0]

'Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a week Jan-June/July then variable in the off season. No limit on PTO - took ~5 weeks last year. One paid conference a year (registration/travel/ 1/2 hotel/ transportation) or online CE program. All licensures & professional group fees covered. Cell phone allowance and mileage reimbursement.|Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a week Jan-June/July then variable in the off season. No limit on PTO - took ~5 weeks last year. One paid conference a year (registration/travel/ 1/2 hotel/ transportation) or online CE program. All licensures & professional group fees covered. Cell phone allowance and mileage reimbursement.|Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a wee

In [13]:
# split comments into individual comments
first_comments = user_comment_df["comments"][0].split("|")

# get the number of comments for first user
len(first_comments)

16

In [14]:
# remove repeated comments
unique_comment = []
for comment in first_comments:
    if comment in unique_comment:
        continue
    else:
        unique_comment.append(comment)

In [15]:
print(f"Length of unique comments for first user: {len(unique_comment)}")
print()
print(unique_comment)

Length of unique comments for first user: 1

['Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a week Jan-June/July then variable in the off season. No limit on PTO - took ~5 weeks last year. One paid conference a year (registration/travel/ 1/2 hotel/ transportation) or online CE program. All licensures & professional group fees covered. Cell phone allowance and mileage reimbursement.']


It can be seen that the comment column contains multiple comments separated with "|"

It can also be seen that there are repeated comments

Lets check for missing values

In [16]:
user_comment_df.isna().sum()

username    0
comments    0
dtype: int64

There are no missig values

Let's check if there are duplicate usernames

In [17]:
if user_comment_df["username"].nunique() == user_comment_df.shape[0]:
    print("There are no duplicated usernames")
else:
    print("There are duplicated usernames")

There are no duplicated usernames


Lets explore the second dataframe also

In [18]:
user_info_df.head()

username  isused   subreddit  created_at
0       LoveAGoodTwist    True  Veterinary  2024-05-02
1            drawntage    True  Veterinary  2024-05-02
2           LinkPast84    True  Veterinary  2024-05-02
3  heatthequestforfire    True  Veterinary  2024-05-02
4       Most-Exit-5507    True  Veterinary  2024-05-02

In [19]:
user_info_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8259 entries, 0 to 8258
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   username    8259 non-null   object
 1   isused      8259 non-null   bool  
 2   subreddit   8259 non-null   object
 3   created_at  8259 non-null   object
dtypes: bool(1), object(3)
memory usage: 201.8+ KB


From the summary above, we se that there are no missing values as each feature has exactly 8259 values which is total entries in the dataset

Let's check if there are duplicate usernames

In [20]:
if user_info_df["username"].nunique() == user_info_df.shape[0]:
    print("There are no duplicated usernames")
else:
    print("There are duplicated usernames")

There are no duplicated usernames


At this point lets create a function to preprocess the comments

## Data Preprocessing

In [36]:
def remove_web_link(text):
    text_list = text.split("|")
    for i in range(len(text_list)):
        text_list[i] = re.sub(r"http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\\(\\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+",
                              "", text_list[i].strip())
    return " | ".join(text_list)

In [37]:
def remove_directories(text):
    text_list = text.split("|")
    for i in range(len(text_list)):
        text_list[i] = re.sub(r"(/[a-zA-Z0-9_]+)+(/)*(.[a-zA-Z_]+)*",
                              "", text_list[i]).strip()
    return " | ".join(text_list)

In [38]:
def remove_punctuations(text):
    text_list = text.split("|")
    for i in range(len(text_list)):
        text_list[i] = ''.join([l for l in text_list[i] if l not in string.punctuation])
    return " | ".join(text_list)

In [39]:
def remove_non_alphabets(text):
    text_list = text.split("|")
    for i in range(len(text_list)):
        text_list[i] = re.sub(r"[^a-zA-Z ]", "", text_list[i].strip())
    return " | ".join(text_list)

In [40]:
def remove_unneeded_spaces(text):
    text_list = text.split("|")
    for i in range(len(text_list)):
        text_list[i] = re.sub(r"(\s)+", " ", text_list[i].strip())
    return " | ".join(text_list)

In [41]:
def remove_repeated_sentence(text):
    text_list = text.split("|")
    unique_comment = []
    for comment in text_list:
        if comment.strip() in unique_comment:
            continue
        else:
            unique_comment.append(comment.strip())
    return " | ".join(unique_comment)

In [42]:
def nlp_preprocessing(text):
    text = remove_web_link(text)
    text = remove_directories(text)
    text = remove_punctuations(text)
    text = remove_non_alphabets(text)
    text = remove_unneeded_spaces(text)
    text = remove_repeated_sentence(text)
    return text

## Hand Engineering

Lets check out the unique values in the subreddit feature as well as the count of each value

In [21]:
subreddit_count = user_info_df['subreddit'].value_counts()
subreddit_count

subreddit
Veterinary          6170
MysteriumNetwork     967
medicine             409
HeliumNetwork        400
orchid               303
vet                   10
Name: count, dtype: int64

In [22]:
subreddit_list = list(subreddit_count.index)

Lets explore each of this subreddit categories starting from the least (the bottom)

In [23]:
# get the number of vet subscribers that are in the first dataset

# initialize counter
user_count = 0
# create container for vet subcribers also in the first dataframe
vet_subscribers = []

# for each username who is a subcriber of vet
for user in user_info_df[user_info_df['subreddit'] == "vet"]["username"]:
    # if username is found in table1
    if not user_comment_df[user_comment_df["username"] == user].empty:
        # increment counter by 1
        user_count += 1
        # capture the username
        vet_subscribers.append(user)

print("Vet Subreddit Count")
print("Table1: {}".format(subreddit_count["vet"]))
print(f"Table2: {user_count}")

Vet Subreddit Count
Table1: 10
Table2: 9


One of the suncribers of vet is not in the first dataset

At this point it would be better to combine both dataset into one

Lets do that

In [24]:
reddit_user_df = pd.merge(user_comment_df, user_info_df,
                          on="username", how="left")

In [25]:
reddit_user_df.head()

username                                           comments  \
0        LoveAGoodTwist  Female, Kentucky.  4 years out. Work equine on...   
1            wahznooski  As a woman of reproductive age, fuck Texas|As ...   
2  Churro_The_fish_Girl  what makes you want to become a vet?|what make...   
3              abarthch  I see of course there are changing variables, ...   
4            VoodooKing  I have 412+ and faced issues because wireguard...   

   isused         subreddit  created_at  
0    True        Veterinary  2024-05-02  
1    True        Veterinary  2024-05-02  
2    True        Veterinary  2024-05-02  
3    True  MysteriumNetwork  2024-05-02  
4   False  MysteriumNetwork  2024-05-03

In [26]:
reddit_user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3276 entries, 0 to 3275
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   username    3276 non-null   object
 1   comments    3276 non-null   object
 2   isused      3276 non-null   bool  
 3   subreddit   3276 non-null   object
 4   created_at  3276 non-null   object
dtypes: bool(1), object(4)
memory usage: 105.7+ KB


Now let us continue with the subreddits

Starting from the bottom and moving up

In [27]:
subreddit_list

['Veterinary',
 'MysteriumNetwork',
 'medicine',
 'HeliumNetwork',
 'orchid',
 'vet']

Let's check out the comments of the vet subscribers

In [28]:
# get list of vet subscribers
vet_sub_list = reddit_user_df[reddit_user_df["subreddit"] == "vet"]["username"].values

# get the comments made by vet subscribers
vet_sub_comments = reddit_user_df[reddit_user_df["subreddit"] == "vet"]["comments"].values

In [29]:
for i in range(len(vet_sub_list)):
    print(f"{vet_sub_list[i]}: {vet_sub_comments[i]}")

test_vet2: The puppy was brought in for its first round of vaccinations.
test_vet3: The adult horse was treated for laminitis.
test_vet4: The juvenile bird was treated for a wing injury.
test_vet5: The senior cat was brought in for a routine health check-up.
test_vet6: I just performed a neutering procedure on a cat.
test_vet7: The dog’s condition is improving after the deworming treatment.
test_vet8: The X-ray showed a fracture in the bird’s wing.
test_vet9: I prescribed flea prevention medication for the puppy.
test_vet: The horse’s blood test revealed signs of equine infectious anemia.


It would be logical for a practicing veterinarian or anyone whose work is related to veterinary to follow "vet" subreddit. This shows closeness to veterinary but does not guarantee being a veterinarian

It can be seen that all these people have just one comment each

Many spoke in 3rd person form which is hard to say if they are doctors or not

Only test_vet6 and test_vet9 can be confirmed to be practicing veterinarian
Others can be classified in the others category since there must be a solid evidence in order to classifier a user as a practicing veterinarian

In [30]:
# store usernames of confirmed veterinarians
veterinarians = ["test_vet6", "test_vet9"]

> It is noteworthy that this type of problem is usually solved effectively with labelled dataset

> However with unlabelled data as the one here, hand engineering may be employed to some extent enough to build a model, therafter the model can predict the rest

> That is the approach I wish to employ for this task

Now, unto next subreddit (orchid)

In [34]:
# get list of orchid subscribers
orchid_sub_list = reddit_user_df[reddit_user_df["subreddit"] == "orchid"]["username"].values

# get the comments made by orchid subscribers
orchid_sub_comment_list = reddit_user_df[reddit_user_df["subreddit"] == "orchid"]["comments"]

In [35]:
print(f"Orchid sunscrbers: {orchid_sub_list}")
print(f"Number of orchid subscribers: {len(orchid_sub_list)}")

Orchid sunscrbers: ['Think_Not_Doer' 'Personal-Escape4283' 'timee_bot' 'erlingspaulsen']
Number of orchid subscribers: 4


It would be neccesary to preprocess our data first as the comments of users who subscribe to orchid is a bit much

In [43]:
orchid_sub_comment_list = orchid_sub_comment_list.apply(nlp_preprocessing)

In [44]:
orchid_sub_comment_list.index

Index([205, 241, 306, 1547], dtype='int64')

In [45]:
orchid_sub_comment_list[1547]

'Nice If anyone is hosting a Mysterium node and have a spare hard disk its very easy to run a Storj node alongside Mysterium I do this on my Raspberry Pi | My Raspberry Pi is running the Mysterium Rasbian image Then I just followed the Storj installation guide and installed Docker engine downloaded their Docker image and started the Storj storage node inside its own container I just looked up Presearch and it looks like you can run their node through a Docker container so this should be okay to do alongside Mysterium To install the Docker engine on Rasbian you can follow this guideinstallusingtheconveniencescriptThanks for bringing it up as I was not aware of this project Might as well try this myself | Doesnt look like Presearch node is supported on Raspberry Pi yet Hardware Specification Docker image is only for x x based CPUs for now We plan to support ARM Raspberry Pi in the future but were currently focused on fixing bugs and making sure the current platform is robust before addit

Checking the comments of all the 4 people who subscribed to orchid shows none of them is either a practicing doctor or a practicing veterinarian

Next subreddit is HeliumNetwork

In [46]:
# get list of orchid subscribers
HeliumNetwork_sub_list = reddit_user_df[reddit_user_df["subreddit"] == "HeliumNetwork"]["username"].values

# get the comments made by orchid subscribers
HeliumNetwork_sub_comment_list = reddit_user_df[reddit_user_df["subreddit"] == "HeliumNetwork"]["comments"]

In [47]:
print(f"HeliumNetwork sunscrbers: {HeliumNetwork_sub_list}")
print(f"Number of HeliumNetwork subscribers: {len(HeliumNetwork_sub_list)}")

HeliumNetwork sunscrbers: ['Best_Bid_9327' 'SoulReaver-SS' 'Imthefatboyb' 'Passi-RVN' 'alexiskef'
 'drunknfoo']
Number of HeliumNetwork subscribers: 6


Thereare 6 sunscribers

Lets proprocess the data of all users who subscribe to HeliumNetwork in order to view it properly

In [48]:
HeliumNetwork_sub_comment_list = HeliumNetwork_sub_comment_list.apply(nlp_preprocessing)

In [49]:
HeliumNetwork_sub_comment_list.index

Index([93, 442, 458, 632, 670, 1504], dtype='int64')

In [50]:
HeliumNetwork_sub_comment_list[1504]

'Was averaging and MYST After three days of mainnet total about MYST and the other node nothing Not worth bothering for me anymore until something changes'

Everone who subscribes to HeliumNetwork belongs to the others category (None of them is perceved to be a practicing doctor or veterinarian)

Lets take alook at the medicine subreddit

In [51]:
# get list of orchid subscribers
medicine_sub_list = reddit_user_df[reddit_user_df["subreddit"] == "medicine"]["username"].values

# get the comments made by orchid subscribers
medicine_sub_comment_list = reddit_user_df[reddit_user_df["subreddit"] == "medicine"]["comments"]

In [52]:
print(f"medicine subscrbers: {medicine_sub_list}")
print(f"Number of medicine subscribers: {len(medicine_sub_list)}")

medicine subscrbers: ['test_doctor2' 'test_doctor3' 'test_doctor4' 'test_doctor5'
 'test_doctor6' 'test_doctor7' 'test_doctor8' 'test_doctor1']
Number of medicine subscribers: 8


Thereare 8 sunscribers

Lets proprocess the data of all users who subscribe to medicine in order to view it properly

In [53]:
medicine_sub_comment_list = medicine_sub_comment_list.apply(nlp_preprocessing)

In [58]:
for i in range(len(medicine_sub_list)):
    print(f"{medicine_sub_list[i]}: {medicine_sub_comment_list.values[i]}")

test_doctor2: The elderly man is recovering from hip replacement surgery
test_doctor3: The teenage boy was treated for a sports injury
test_doctor4: The woman is expecting a baby and visited for a prenatal checkup
test_doctor5: I just performed an appendectomy on a patient
test_doctor6: The patients blood pressure is stabilizing after the medication
test_doctor7: The MRI scan revealed a tumor in the patients brain
test_doctor8: I prescribed antibiotics for the patients bacterial infection
test_doctor1: The patients EKG showed signs of a possible heart attack


Just like the vet subreddit, most of the users in this category speak in third person form which makes it hard to say if they are really practicing doctor or a medical practitioner like nurse student etc

Only test_doctor5 and tes_doctor8 are confirmed to be practicing doctors

In [59]:
Doctors = ["test_doctor5", "test_doctor8"]
Doctors

['test_doctor5', 'test_doctor8']

Lets take alook at the next subreddit, MysteriumNetwork

In [60]:
# get list of MysteriumNetwork subscribers
MysteriumNetwork_sub_list = reddit_user_df[reddit_user_df["subreddit"] == "MysteriumNetwork"]["username"].values

# get the comments made by MysteriumNetwork subscribers
MysteriumNetwork_sub_comment_list = reddit_user_df[reddit_user_df["subreddit"] == "MysteriumNetwork"]["comments"]

In [62]:
print(f"Number of MysteriumNetwork subscribers: {len(MysteriumNetwork_sub_list)}")

Number of MysteriumNetwork subscribers: 967


There are quite a lot of users who subscribe to MysteriumNetwork

Nevertheless we still need to preprocess the comments

In [65]:
MysteriumNetwork_sub_comment_list = MysteriumNetwork_sub_comment_list.apply(nlp_preprocessing)

Lets take alook at the last subreddit also, Veterinary

In [63]:
# get list of Veterinary subscribers
Veterinary_sub_list = reddit_user_df[reddit_user_df["subreddit"] == "Veterinary"]["username"].values

# get the comments made by Veterinary subscribers
Veterinary_sub_comment_list = reddit_user_df[reddit_user_df["subreddit"] == "Veterinary"]["comments"]

In [64]:
print(f"Number of Veterinary subscribers: {len(Veterinary_sub_list)}")

Number of Veterinary subscribers: 2282


There are quite a lot of users who subscribe to Veterinary also

Lets first preprocess the comments

In [66]:
Veterinary_sub_comment_list = Veterinary_sub_comment_list.apply(nlp_preprocessing)

At this point is would be better to build the model on the 4 subreddit we have checked so far then predict the category for the remianing users who subscribe to the remaining 2 categories

## Model Building